File: chapitre7/exercice7.3.py
import os
import signal
import sys
import time

# Nombre de fils à créer [1, 2]
NB_FILS = 5
# Tableau pour stocker les PID des fils [3, 4]
pids =  * NB_FILS
# Variable pour aider le gestionnaire à identifier le rôle (Père ou Fils)
est_parent = True

def handle_sigterm(sig, frame):
    """Gestionnaire de signal pour le message de fin [5, 6]"""
    self_pid = os.getpid() # [3, 7]
    if not est_parent:
        print(f"[Fils {self_pid}] meurs")
        sys.exit(0)
    else:
        print(f"[Père {self_pid}] tué->Fin du programme.")
        sys.exit(0)

def fils_process(index):
    """Fonction exécutée par chaque fils [3, 8]"""
    global est_parent
    est_parent = False
    
    # Chaque fils installe son gestionnaire pour SIGTERM [6, 9]
    signal.signal(signal.SIGTERM, handle_sigterm)
    
    print(f"Fils {index} (PID: {os.getpid()}) démarré.")
    while True:
        # Boucle infinie demandée [2]
        time.sleep(1) 

def main():
    global pids
    # Installation du gestionnaire pour le père [6]
    signal.signal(signal.SIGTERM, handle_sigterm)

    # Création des 5 fils via fork() [1, 8]
    for i in range(NB_FILS):
        pid = os.fork()
        if pid == 0:
            # Code du processus fils [3]
            fils_process(i)
            sys.exit(0)
        else:
            # Code du processus père : on stocke le PID du fils [3, 4]
            pids[i] = pid

    # Boucle de menu pour le père [2]
    while True:
        try:
            print(f"\n--- MENU PERE (Mon PID: {os.getpid()}) ---")
            print("1. Endormir un fils (SIGSTOP)")
            print("2. Réveiller un fils (SIGCONT)")
            print("3. Terminer un fils (SIGTERM)")
            print("4. Quitter (Tue le père et les fils)")
            
            choix_input = input("Choix : ")
            if not choix_input: continue
            choix = int(choix_input)

            if choix == 4:
                # Le père termine tous ses fils avant de se tuer lui-même [4, 10]
                for i in range(NB_FILS):
                    try:
                        os.kill(pids[i], signal.SIGTERM)
                    except ProcessLookupError:
                        pass
                os.kill(os.getpid(), signal.SIGTERM)

            num_fils_input = input(f"Numéro du fils (0 à {NB_FILS - 1}) : ")
            if not num_fils_input: continue
            num_fils = int(num_fils_input)

            if num_fils < 0 or num_fils >= NB_FILS:
                print("Fils invalide.")
                continue

            target_pid = pids[num_fils]

            # Envoi des signaux selon le choix [4, 10]
            if choix == 1:
                os.kill(target_pid, signal.SIGSTOP) # Suspendre [11]
                print(f"Fils {num_fils} suspendu.")
            elif choix == 2:
                os.kill(target_pid, signal.SIGCONT) # Reprendre [11]
                print(f"Fils {num_fils} réveillé.")
            elif choix == 3:
                os.kill(target_pid, signal.SIGTERM) # Terminer [9]
                print(f"Demande de terminaison envoyée au fils {num_fils}.")
            else:
                print("Option inconnue.")
                
        except (ValueError, EOFError):
            print("\nEntrée invalide ou fin de programme.")
            break

if __name__ == "__main__":
    main()

File: chapitre7/exercice7.3.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/wait.h>

#define NB_FILS 5

pid_t pids[NB_FILS];

// Gestionnaire de signal pour le père et les fils
void handle_sigterm(int sig) {
    pid_t self = getpid();
    int is_parent = 0;
    
    for(int i = 0; i < NB_FILS; i++) {
        if (pids[i] == 0) { // On est dans un fils (approximatif ici, mieux vaut vérifier via getppid)
             printf("[Fils %d] meurs\n", self);
             exit(0);
        }
    }
    // Si le code arrive ici, c'est le père
    printf("[Père %d] tué->Fin du programme.\n", self);
    exit(0);
}

void fils_process(int index) {
    // Chaque fils installe son propre gestionnaire pour le message de fin
    signal(SIGTERM, handle_sigterm);
    
    printf("Fils %d (PID: %d) démarré.\n", index, getpid());
    while(1) {
        // Boucle infinie demandée
        sleep(1); 
    }
}

int main() {
    int choix, num_fils;

    // Installation du gestionnaire pour le père
    signal(SIGTERM, handle_sigterm);

    // Création des 5 fils
    for (int i = 0; i < NB_FILS; i++) {
        pids[i] = fork();
        if (pids[i] == 0) {
            fils_process(i);
            exit(0); 
        }
    }

    // Boucle de menu pour le père
    while (1) {
        printf("\n--- MENU PERE (Mon PID: %d) ---\n", getpid());
        printf("1. Endormir un fils (SIGSTOP)\n");
        printf("2. Réveiller un fils (SIGCONT)\n");
        printf("3. Terminer un fils (SIGTERM)\n");
        printf("4. Quitter (Tue le père et les fils)\n");
        printf("Choix : ");
        scanf("%d", &choix);

        if (choix == 4) {
            for(int i=0; i<NB_FILS; i++) kill(pids[i], SIGTERM);
            raise(SIGTERM); // Le père s'envoie le signal à lui-même
        }

        printf("Numéro du fils (0 à %d) : ", NB_FILS - 1);
        scanf("%d", &num_fils);

        if (num_fils < 0 || num_fils >= NB_FILS) {
            printf("Fils invalide.\n");
            continue;
        }

        switch (choix) {
            case 1:
                kill(pids[num_fils], SIGSTOP);
                printf("Fils %d suspendu.\n", num_fils);
                break;
            case 2:
                kill(pids[num_fils], SIGCONT);
                printf("Fils %d réveillé.\n", num_fils);
                break;
            case 3:
                kill(pids[num_fils], SIGTERM);
                printf("Demande de terminaison envoyée au fils %d.\n", num_fils);
                break;
            default:
                printf("Option inconnue.\n");
        }
    }

    return 0;
}